File: app/src/main/java/com/sdex/activityrunner/glide/ActivityIconModelLoader.java
package com.sdex.activityrunner.glide;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.bumptech.glide.Priority;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.Options;
import com.bumptech.glide.load.data.DataFetcher;
import com.bumptech.glide.load.model.ModelLoader;
import com.bumptech.glide.signature.ObjectKey;
import com.sdex.activityrunner.db.activity.ActivityModel;

class ActivityIconModelLoader implements ModelLoader<ActivityModel, ActivityModel> {

  @Nullable
  @Override
  public LoadData<ActivityModel> buildLoadData(@NonNull final ActivityModel applicationInfo,
                                               int width, int height, @NonNull Options options) {
    return new LoadData<>(new ObjectKey(applicationInfo), new DataFetcher<ActivityModel>() {
      @Override
      public void loadData(@NonNull Priority priority,
                           @NonNull DataFetcher.DataCallback<? super ActivityModel> callback) {
        callback.onDataReady(applicationInfo);
      }

      @Override
      public void cleanup() {

      }

      @Override
      public void cancel() {

      }

      @NonNull
      @Override
      public Class<ActivityModel> getDataClass() {
        return ActivityModel.class;
      }

      @NonNull
      @Override
      public DataSource getDataSource() {
        return DataSource.LOCAL;
      }
    });
  }

  @Override
  public boolean handles(@NonNull ActivityModel applicationInfo) {
    return true;
  }
}

File: app/src/main/java/com/sdex/activityrunner/glide/ActivityIconModelLoader.kt
package com.sdex.activityrunner.glide

import com.bumptech.glide.Priority
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.Options
import com.bumptech.glide.load.data.DataFetcher
import com.bumptech.glide.load.model.ModelLoader
import com.bumptech.glide.signature.ObjectKey
import com.sdex.activityrunner.app.ActivityModel

internal class ActivityIconModelLoader : ModelLoader<ActivityModel, ActivityModel> {

    override fun buildLoadData(
        applicationInfo: ActivityModel,
        width: Int, height: Int, options: Options
    ): ModelLoader.LoadData<ActivityModel> {
        return ModelLoader.LoadData(
            ObjectKey(applicationInfo),
            object : DataFetcher<ActivityModel> {
                override fun loadData(
                    priority: Priority,
                    callback: DataFetcher.DataCallback<in ActivityModel>
                ) {
                    callback.onDataReady(applicationInfo)
                }

                override fun cleanup() {

                }

                override fun cancel() {

                }

                override fun getDataClass(): Class<ActivityModel> {
                    return ActivityModel::class.java
                }

                override fun getDataSource(): DataSource {
                    return DataSource.LOCAL
                }
            })
    }

    override fun handles(applicationInfo: ActivityModel) = true
}
